File: dynamic_programming/double_linked_list.cpp
using namespace std;
#include <iostream>

class Node {
public:
    int data;
    Node* next;
    Node* prev;
    Node(int data) {
        this->data = data;
        this->next = NULL;
        this->prev = NULL;
    }
};

class DoublyLinkedList {
public:
    Node* head;
    DoublyLinkedList() {
        this->head = NULL;
    }

    void add(int data) {
        Node* new_node = new Node(data);
        if (this->head == NULL) {
            this->head = new_node;
            return;
        }
        Node* current = this->head;
        while (current->next != NULL) {
            current = current->next;
        }
        current->next = new_node;
        new_node->prev = current;
    }

    Node* search(int data) {
        Node* current = this->head;
        while (current != NULL) {
            if (current->data == data) {
                return current;
            }
            current = current->next;
        }
        return NULL;
    }

    void insert_after(Node* node, int data) {
        Node* new_node = new Node(data);
        new_node->next = node->next;
        if (node->next != NULL) {
            node->next->prev = new_node;
        }
        node->next = new_node;
        new_node->prev = node;
    }

    void delete_node(int data) {
        Node* current = this->head;
        if (current->data == data) {
            this->head = current->next;
            if (current->next != NULL) {
                current->next->prev = NULL;
            }
            delete current;
            return;
        }
        while (current != NULL) {
            if (current->next->data == data) {
                Node* temp = current->next;
                current->next = current->next->next;
                if (current->next != NULL) {
                    current->next->prev = current;
                }
                delete temp;
                return;
            }
            current = current->next;
        }
    }

    void reverse() {
        Node* prev = NULL;
        Node* current = this->head;
        Node* next = NULL;
        while (current != NULL) {
            next = current->next;
            current->next = prev;
            //Check if prev is Null
            if (prev != NULL) {
                prev->prev = current;
            }
            prev = current;
            current = next;
        }
        this->head = prev;
    }

    void rotate(int k) {
        Node* current = this->head;
        int n = 0;
        while (current != NULL) {
            n++;
            current = current->next;
        }
        k = k % n;
        if (k == 0) {
            return;
        }
        Node* slow = this->head;
        Node* fast = this->head;
        for (int i = 0; i < k; i++) {
            fast = fast->next;
        }
        while (fast->next != NULL) {
            slow = slow->next;
            fast = fast->next;
        }
        fast->next = this->head;
        this->head->prev = fast;
        this->head = slow->next;
        slow->next = NULL;

    }

    void print() {
        Node* current = this->head;
        while (current != NULL) {
            cout << current->data << " ";
            current = current->next;
        }
        cout << endl;
    }
};

int main() {
    DoublyLinkedList dl;
    dl.add(1);
    dl.add(2);
    dl.add(3);
    dl.add(4);
    dl.add(5);
    dl.print();
    dl.insert_after(dl.search(3), 9);
    dl.print();
    dl.delete_node(3);
    dl.print();
    dl.reverse();
    dl.print();
    dl.rotate(2);
    dl.print();

    return 0;
}

File: dynamic_programming/double_linked_list.java
package dynamic_programming;

class Node {
    int data;
    Node next;
    Node prev;

    Node(int data) {
        this.data = data;
        this.next = null;
        this.prev = null;
    }
}

class DoublyLinkedList {
    Node head;

    DoublyLinkedList() {
        this.head = null;
    }

    public void addNode(int data) {
        Node newNode = new Node(data);

        if (head == null) {
            head = newNode;
        } else {
            Node temp = head;
            while (temp.next != null) {
                temp = temp.next;
            }
            temp.next = newNode;
            newNode.prev = temp;
        }
    }

    Node searchNode(int data) {
        Node current = head;
        while (current != null) {
            if (current.data == data) {
                return current;
            }
            current = current.next;
        }
        return null;
    }

    public void insertAfter(Node prevNode, int data) {
        if (prevNode == null) {
            System.out.println("The given previos node cannot be null");
            return;
        }

        Node newNode = new Node(data);
        newNode.next = prevNode.next;
        newNode.prev = prevNode;
        prevNode.next = newNode;
    }

    public void deleteNode(int data) {
        Node current = head;
        if (current.data == data) {
            head = current.next;
            if (current.next != null) {
                current.next.prev = null;
            }
            return;
        }

        while (current != null) {
            if (current.next.data == data) {
                current.next = current.next.next;
                if (current.next != null) {
                    current.next.prev = current;
                }
                return;
            }
            current = current.next;
        }
    }

    public void reverse() {
        Node prev = null;
        Node current = head;
        Node next = null;
        while (current != null) {
            next = current.next;
            current.next = prev;
            if (prev != null) {
                prev.prev = current; // Fix: Assign 'prev' to 'current' node's 'prev' pointer instead of 'next' pointer
            }
            prev = current;
            current = next;
        }
        head = prev;
    }

    void rotate(int k) {
        Node currentNode = head;
        int n = 0;
        while (currentNode != null) {
            n = n + 1;
            currentNode = currentNode.next;
        }
        k = k % n;
        if (k == 0) {
            return;
        }
        Node slow = head;
        Node fast = head;
        for (int i = 0; i < k; i++) {
            fast = fast.next;
        }
        while (fast.next != null) {
            slow = slow.next;
            fast = fast.next;
        }
        fast.next = head;
        head.prev = fast;
        head = slow.next;
        slow.next = null;
    }

    public void printList() {
        Node current = head;
        while (current != null) {
            System.out.print(current.data + " ");
            current = current.next;
        }
        System.out.println();
    }
}


public class double_linked_list {
    public static void main(String[] args) {
        DoublyLinkedList dl = new DoublyLinkedList();
        
        dl.addNode(1);
        dl.addNode(2);
        dl.addNode(3);
        dl.addNode(4);
        dl.addNode(5);
        dl.printList();
        dl.insertAfter(dl.searchNode(3), 9);
        dl.printList();
        dl.deleteNode(9);
        dl.printList();
        dl.reverse();
        dl.printList();
        dl.rotate(3);
        dl.printList();
    }
    
}
